### PR TITLE
Updates PR #627 - Suffix Arrays

### PR DESCRIPTION
**Pull request by pierce314159**
This pull request updates PR #627 submitted by @zhihuidu.

This PR:
- Brings `string-suffix-array-functionality` up to date with master
- Rearranges code to increase separation between `strings` and `suffix arrays` and minimize diff with master/highlight changes
- Removes external `libdivsufsort` in favor of zhihuidu's native chapel suffix array construction implementation
- Changes formatting in some areas
___________________________

<details>
  <summary>bf1757c Repackages zhihuidu's commits in PR#627</summary>

- Merge squashed commits and rebased over a fresh copy of master. This makes the commits more portable and the git history of master more readable with all commits grouped together
- Helps minimize diff with master since I default to master in conflicts where there are only formatting differences
</details>

<details>
  <summary>07f51d8 Applies original minimum changes to resolve conflicts with master</summary>

- Updates logging statements to use logLevel instead of v 
  - PR #760
- Updates `check` to `check_table` 
  - PR #792
- Removes outdated `Strings.attach` function which used `_` for subcomponents (new convention is `.` )
  - PR #774  
- Makes `unregister_strings_by_name` a staticmethod
- Applies recent changes to `SegmentedArray.chpl` (automatic merge was thrown off by addition of `class SegSArray`) including:
  - PR #666
  - PR #781
</details>

<details>
  <summary>ca4941b Implements changes suggested in PR 627 feedback</summary>

- Minimizes diff with master to highlight functionality changes
- Moves `SArrarys` class to `suffix_array.py`
- Refomats `suffix_array.py`
- Reformats `SACA.chpl` to fix bracket alignment
- Removes commented out code
- Adds type hints to size and bytes in `strings.py` for mypy
- Replaces `writelns` in `SegmentedArray.chpl` with logging statements following the example of `SegString` class

Previously stand-alone commits have been squashed into this one since Zhihui approved them
- Removes external [libdivsufsort](https://github.com/y-256/libdivsufsort) code and references. Defaults to zhihuidu's native Chapel suffix array construction implementation using skew algorithm (based on [Simple Linear Work Suffix Array Construction](https://www.cs.helsinki.fi/u/tpkarkka/publications/jacm05-revised.pdf))
 - Using the native chapel suffix array construction implementation prevents copying the `libdivsufsort` codebase into arkouda or forcing the dependency on arkouda users.
 - Renames `class SegSArray` to `class SegSuffixArray` and moves this class to `SegmentedSuffixArray.chpl` (previously in `SegmentedArray.chpl` along with `SegString`). Rearranged to increase separation with `strings` and minimize diff with master. 
</details>

<details>
  <summary>935f29b Implements changes suggested in this PR's feedback</summary>

- Removes unused `fields` variable in `_parse_single_int_array_value`
- Passes `Exception` into `ValueError` in `_parse_single_int_array_value`
- Inverts logic of `in1d_int` to avoid duplicating the `ind1d` logic
- Adds `self.name` to `SArray` class and updates attach/register functions to use dots instead of dashs to be consistent with the `Strings` class
- Removes unnecessary cast in `suffix_array_file`
- Changes typing from `sipHash128` and `computeSipHashLocalized` to accept `int` and `uint(8)`
- Updates `suffix_array` registration and info functions
- Renames `SACA.chpl` to `SuffixArrayConstruction.chpl`
- Adds `UnitTestSuffixArrayConstruction.chpl`
- Adds docstrings to `SuffixArrayConstruction.chpl` and removes duplicated logic by modifying `radixPass` to accept `int` or `uint(8)`
</details>

### CODE REVIEW

**Comment by pierce314159** [comment]
@zhihuidu, are you okay with the changes I've made here? I specifically made the move/rename `class SegSArray` and remove `libdivsufsort` commits stand alone, so they are easy to revert

Once I have your approval, I'll invite the other core team members to review

File: arkouda/suffix_array.py
Comment (glitch) [inline arkouda/suffix_array.py:468]: I don't think you need the explicit cast to str here since it was already done to `tmpmsg`

File: arkouda/suffix_array.py
Comment (glitch) [inline arkouda/suffix_array.py:103]: Note, if we want to keep this in line with `Strings` then we should add
`self.name: Optional[str] = None`  here along with updates to the register/attach, etc. methods.

Note, the attach/register suite of methods now use a form of `user_defined_name + ".offsets/bytes"` etc.  where we now use a dot character instead of an underscore.

It doesn't necessarily need to be changed now but it should be immediate follow-on work to be done ASAP.

File: arkouda/pdarrayclass.py
Comment (hokiegeek2) [inline arkouda/pdarrayclass.py:101]: I recommend passing in the Exception as e into ValueError to get more context for debug purposes

File: arkouda/suffix_array.py
Comment (hokiegeek2) [reply arkouda/suffix_array.py:468]: Agreed, good catch

File: arkouda/suffix_array.py
Comment (hokiegeek2) [inline arkouda/suffix_array.py:51]: I recommend replacing the objtype string with type argument or Enum

File: src/SACA.chpl
Comment (glitch) [inline src/SACA.chpl:17]: Additional parenthesis should be added to the logic statements above for clarity.

File: arkouda/suffix_array.py
Comment (hokiegeek2) [reply arkouda/suffix_array.py:103]: 👍

File: arkouda/suffix_array.py
Comment (hokiegeek2) [inline arkouda/suffix_array.py:120]: why is this self.size - 3?

**Review by hokiegeek2 — APPROVED** [review]

File: src/SACA.chpl
Comment (glitch) [inline src/SACA.chpl:207]: General note: There is a lot going on in here, it would be nice to get
1. Chapel doc comments
2. Chapel based unit tests to illustrate proper function.

File: src/SipHash.chpl
Comment (glitch) [inline src/SipHash.chpl:131]: It's unfortunate the proc needed to be copied just to accept a different type arg.  I wonder if the typing was removed if Chapel would be happy?

File: arkouda/suffix_array.py
Comment (pierce314159) [reply arkouda/suffix_array.py:51]: That makes sense to me. I think this is just following along with what is done in `Strings` and `pdarrayclass` where the `objtype` is a literal string. Would it make sense to hold off on this change and update along with the others in a separate ticket?

https://github.com/Bears-R-Us/arkouda/blob/b83c3ba320ef7f64cc3521b487f4efbea88b162a/arkouda/pdarrayclass.py#L99

https://github.com/Bears-R-Us/arkouda/blob/b83c3ba320ef7f64cc3521b487f4efbea88b162a/arkouda/strings.py#L52

File: src/SipHash.chpl
Comment (pierce314159) [reply src/SipHash.chpl]: It worked! Good call

File: src/SACA.chpl
Comment (pierce314159) [reply src/SACA.chpl:207]: @zhihuidu do you mind adding the chapel docs and unit tests for `SACA.chpl` since you're more familiar with this section than me?

Also if you could add the parenthesis to clarify the logic in this section like glitch suggested above while you're at it, that'd be great

```python
  inline proc leq(a1:int, a2:int, b1:int, b2:int) {
    // lexicographic order
    return(a1 < b1 || a1 == b1 && a2 <= b2);
  } // for pairs

  inline proc leq(a1:int, a2:int, a3:int, b1:int, b2:int, b3:int) {
    // lexicographic order
    return(a1 < b1 || a1 == b1 && leq(a2, a3, b2, b3));
  } // for pairs
```

You can pull down the latest in my `suffix_array` branch and make a pull request with these changes?

File: arkouda/suffix_array.py
Comment (hokiegeek2) [reply arkouda/suffix_array.py:51]: @pierce314159 yeah, good idea to work that in a separate issue

File: arkouda/suffix_array.py
Comment (hokiegeek2) [reply arkouda/suffix_array.py:120]: @pierce314159 gotcha, okay

File: src/SipHash.chpl
Comment (ronawho) [reply src/SipHash.chpl]: Note that you should be able to use `integral` to cover both `int` and `uint(8)` without going totally generic.

Alternatively if you want to just restrict to int/uint(8) you should be able to do something like:

```chpl
private proc computeSipHashLocalized(msg: [] ?t, D, param outlen: int) where t == int || t == uint(8) {
```

Comment (glitch) [reply src/SipHash.chpl]: @ronawho  Oh.. that's cool!

**Review by glitch — APPROVED** [review]

File: src/SACA.chpl
Comment (pierce314159) [reply src/SACA.chpl]: I've added parenthesis in both of the `leq` procs and renamed them for clarity

File: src/SACA.chpl
Comment (pierce314159) [reply src/SACA.chpl]: I added `UnitTestSuffixArrayConstruction.chpl` with expected output in `UnitTestSuffixArrayConstruction.good`. This covers both the example from the paper and from client-side `suffix-array` docstring. 

I added docstrings and comments to the best of my ability using `Section 3` and `appendix A` of [Linear Work Suffix Array Construction](https://www.cs.helsinki.fi/u/tpkarkka/publications/jacm05-revised.pdf). These can certainly be improved and this section rewritten for clarity because it is still quite busy and difficult to follow.

File: src/SuffixArrayConstruction.chpl
Comment (pierce314159) [inline src/SuffixArrayConstruction.chpl:56]: I was able to use @ronawho's type declaration trick again to modify  `radixPass` to accept `int` or `uint(8)` eliminating a need for a second function!

**Comment by pierce314159** [comment]
@zhihuidu, I rebased over master so the `suffix_array` branch is up to date with no merge conflicts. I also made some slight changes to `SuffixArrayConstruction` to shorten the code. I verified the functionality remained the same with `python3 benchmarks/sa.py localhost 5555 --correctness-only`

**Comment by zhihuidu** [comment]
> @zhihuidu, I rebased over master so the `suffix_array` branch is up to date with no merge conflicts. I also made some slight changes to `SuffixArrayConstruction` to shorten the code. I verified the functionality remained the same with `python3 benchmarks/sa.py localhost 5555 --correctness-only`

@pierce314159 I appreciate it. In fact, you almost rewrite the code.

**Comment by mhmerrill** [comment]
@pierce314159 @zhihuidu I would like to see this updated and moved to the `arkouda-njit` or `arkouda-contrib` so it can be included as an optional module in the arkouda build.

**Comment by zhihuidu** [comment]
@mhmerrill @pierce314159 No problem. I can rewrite the suffix array code based on Kyle's framework to integrate new functionalities and push it to arkouda-njit first.

**Comment by pierce314159** [comment]
> @mhmerrill @pierce314159 No problem. I can rewrite the suffix array code based on Kyle's framework to integrate new functionalities and push it to arkouda-njit first.

It might be easier than that, let me take a crack at it and see if I can use git magic to get it in line with the `njit` repo

**Comment by zhihuidu** [comment]
@pierce314159 If some tools can do it automatically, it will be better. Please go ahead.

**Comment by glitch** [comment]
> > @mhmerrill @pierce314159 No problem. I can rewrite the suffix array code based on Kyle's framework to integrate new functionalities and push it to arkouda-njit first.
> 
> @zhihuidu It might be easier than that, let me take a crack at it and see if I can use git magic to get it in line with the `njit` repo

Note: It looked like @pierce314159 didn't have access to the repo, that should be fixed now. :+1:

**Pull request closed without merging**
